modules/fedimint-ln-client/src/lib.rs

```diff
@@ -1279,49 +1221,65 @@ impl LightningClientModule {
 
                 yield LnPayState::Created;
 
-                if client_ctx
-                    .transaction_updates(operation_id)
-                    .await
-                    .await_tx_accepted(out_point.txid)
-                    .await
-                    .is_err()
-                {
-                    yield LnPayState::Canceled;
-                    return;
-                }
-                yield LnPayState::Funded;
-
-                match self_ref.await_lightning_payment_success(operation_id).await {
-                    Ok(preimage) => {
-                        if !change.is_empty() {
-                            yield LnPayState::AwaitingChange;
-                            match client_ctx.await_primary_module_outputs(operation_id, change).await {
-                                Ok(_) => {}
-                                Err(_) => {
-                                    yield LnPayState::UnexpectedError { error_message: "Error occurred while waiting for the primary module's output".to_string() };
-                                    return;
+                let mut stream = self_ref.notifier.subscribe(operation_id).await;
+
+                loop {
+                    match stream.next().await {
+                        Some(LightningClientStateMachines::LightningPay(state)) => match state.state {
```

Comment:
This being a `loop` makes me slightly nervous. This means we don't enforce the correct order of events anymore and might miss implementation bugs that way. I think fixing the bug that prompted this PR is more important and I'm thinking about an abstraction that lets us model these high-level state graphs with less boilerplate while enforcing all invariants.

Reply:
I also don't like the loop here. I re-wrote this without a loop that enforces ordering. Hopefully this is more intuitive.